be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp

```diff
@@ -84,8 +127,10 @@ Status LocalExchangeSourceOperator::set_finished(RuntimeState* state) {
 
 StatusOr<ChunkPtr> LocalExchangeSourceOperator::pull_chunk(RuntimeState* state) {
     ChunkPtr chunk = _pull_passthrough_chunk(state);
-    if (chunk == nullptr) {
+    if (chunk == nullptr && _key_partition_pending_chunk_empty()) {
         chunk = _pull_shuffle_chunk(state);
+    } else if (chunk == nullptr && !_key_partition_pending_chunk_empty()) {
+        chunk = _pull_key_partition_chunk(state);
     }
     return std::move(chunk);
```

Comment:
this logic is not intuitive, can we just set a type to mark it's type? shuffle/passthrough/key_partition? So all the logic of `is_finished` and `has_output` can be simplified.